components/file_system/src/rate_limiter.rs

```diff
@@ -1,43 +1,350 @@
+                    let cached_last_refill_time = *last_refill_time;
+                    let (mut last_refill_time, timed_out) = self
+                        .condv
+                        .async_wait_timeout(
+                            &self.last_refill_time,
+                            last_refill_time,
+                            self.refill_period - since_last_refill,
+                        )
+                        .await;
+                    let now = Instant::now();
+                    if timed_out && *last_refill_time == cached_last_refill_time {
+                        *last_refill_time = now;
+                        break self.refill_and_request(bytes);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Default for PerTypeIORateLimiter {
+    fn default() -> PerTypeIORateLimiter {
+        PerTypeIORateLimiter::new(0, Duration::from_millis(10))
+    }
+}
 
-/// No-op limiter
 /// An instance of `IORateLimiter` should be safely shared between threads.
 #[derive(Debug)]
 pub struct IORateLimiter {
-    refill_bytes: atomic::AtomicUsize,
+    write_limiters: [PerTypeIORateLimiter; IO_TYPE_VARIANTS],
+    read_limiters: [PerTypeIORateLimiter; IO_TYPE_VARIANTS],
+    total_limiters: [PerTypeIORateLimiter; IO_TYPE_VARIANTS],
+    recorder: Arc<BytesRecorder>,
 }
 
 impl IORateLimiter {
-    pub fn new(_refill_bytes: usize) -> IORateLimiter {
-        IORateLimiter {
-            refill_bytes: atomic::AtomicUsize::new(_refill_bytes),
+    // TODO: pass in rate limiting options
+    pub fn new(bytes_per_sec: usize, recorder: Arc<BytesRecorder>) -> IORateLimiter {
+        let limiter = IORateLimiter {
+            write_limiters: Default::default(),
+            read_limiters: Default::default(),
+            total_limiters: Default::default(),
+            recorder,
+        };
+        if bytes_per_sec != 0 {
+            for l in limiter.write_limiters.iter() {
+                l.set_bytes_per_sec(bytes_per_sec);
+            }
+            for l in limiter.read_limiters.iter() {
+                l.set_bytes_per_sec(bytes_per_sec);
+            }
+            for l in limiter.total_limiters.iter() {
+                l.set_bytes_per_sec(bytes_per_sec);
+            }
         }
+        limiter
     }
 
-    /// Request for token for bytes and potentially update statistics. If this
+    /// Requests for token for bytes and potentially update statistics. If this
     /// request can not be satisfied, the call is blocked. Granted token can be
     /// less than the requested bytes, but must be greater than zero.
-    pub fn request(&self, _io_type: IOType, _io_op: IOOp, bytes: usize) -> usize {
-        std::cmp::min(self.refill_bytes.load(atomic::Ordering::Relaxed), bytes)
+    pub fn request(&self, io_type: IOType, io_op: IOOp, bytes: usize) -> usize {
+        let prio = get_priority(io_type);
+        let mut bytes = self.total_limiters[IOType::Other as usize].request(bytes, prio);
```

Comment:
why use `IOType::Other`

Reply:
Here uses `Other` slot to store total limiter, which is the sum of all r/w IOs. Just for convenience because `Other` itself doesn't own a dedicated limiter.